pkg/promtail/promtail.go

```diff
@@ -4,56 +4,72 @@ import (
 	"sync"
 
 	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log"
 
 	"github.com/grafana/loki/pkg/promtail/client"
 	"github.com/grafana/loki/pkg/promtail/config"
 	"github.com/grafana/loki/pkg/promtail/server"
 	"github.com/grafana/loki/pkg/promtail/targets"
 )
 
+// Option is a function that can be passed to the New method of Promtail and
+// customize the Promtail that is created.
+type Option func(p *Promtail)
+
+// WithLogger overrides the default logger for Promtail.
+func WithLogger(log log.Logger) Option {
```

Comment:
This seems like a bit of a weird way to expose this. Since we're expecting to modify this externally, couldn't we just make the field public? Perhaps you wanted to make it only modifiable via this new function. I'm not really opinionated here though.

Reply:
It'd have to be a public field in the Config struct for this to work since the logger is already passed around when New is called. I'm open to this, but I think it'd be a little strange to do that.

I do agree that the Option pattern is a big heavy handed for this specific PR, but my justification is that there's going to be a second `With<X>` function in my final PR for #2405, where you give it a custom mux router, where having two option functions would feel more justified. 

I don't feel that strongly about this either though, I'm happy to make it a public field of Config if you think that'd be a better approach here.